README.md
# keycloak-scanner

## Introduction

This scanner scan keycloak for known vulnerabilities.


## Installation

```
pip install --upgrade keycloak-scanner
```


## Example
```
$ git clone https://github.com/NeuronAddict/keycloak-scanner
$ cd keycloak-scanner
$ docker-compose -f itests/docker-compose.yml up -d
$ python3 itests/wait-docker-compose.py # just wait keycloak to be load # may be you neeed 'pip install waiting'
python3 itests/wait-docker-compose.py           
('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))
...
HTTPConnectionPool(host='localhost', port=8080): Read timed out. (read timeout=1)
HTTPConnectionPool(host='localhost', port=8080): Read timed out. (read timeout=1)
Keycloak seems to be loaded
$ keycloak-scanner http://localhost:8080 --realms master --clients account --username admin --password Pa55w0rd 
$ # http://localhost:8080  # url to test 
$ #--realms master  # realms to scan, check if a realm exists and use this realms to further scans
$ #--clients account  # clients to scan, check if a client exists and use it to further scans
$ #--username admin  # add a username to test the auth process 
$ #--password Pa55w0rd  # password to test a password auth 
[INFO] Start scanner RealmScanner...
[INFO] Find realm master (http://localhost:8080/auth/realms/master)
[INFO] Public key for realm master : MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyF7ytt1AcJaN67GkLKNrPL6ljoTyYMzMFZ/fXmEJw52yvAXCqE2qFM4MH+fRDfzYcougyOIwNHbDqfAmKzKpeYGi+4JSaSmDGpZVuz2aDkncyXh6uA4IanjBai7IhEeWDY6HCcLxkd/ppfNclmfOrEGJGbFoz+QCFiNbWzSr0mAo1S3WmgC13297nK5iunR+eJSqCbg3FXn+8RZcwhNHhKSGV75G4ZnBDLcBcaEUflBWshv2gAErZktT0tdEtXNRpv4vAvp0yEvAKSPVOESpnZW7PFNtBPI/+GlaAWxEC9V58qzhiRTJ+MU3fzwcBMRz4DmptdSN6bDLvkPr5eS9JQIDAQAB
[INFO] Start scanner WellKnownScanner...
[INFO] Find a well known for realm Realm('master', 'http://localhost:8080/auth/realms/master', {'realm': 'master', 'public_key': 'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAyF7ytt1AcJaN67GkLKNrPL6ljoTyYMzMFZ/fXmEJw52yvAXCqE2qFM4MH+fRDfzYcougyOIwNHbDqfAmKzKpeYGi+4JSaSmDGpZVuz2aDkncyXh6uA4IanjBai7IhEeWDY6HCcLxkd/ppfNclmfOrEGJGbFoz+QCFiNbWzSr0mAo1S3WmgC13297nK5iunR+eJSqCbg3FXn+8RZcwhNHhKSGV75G4ZnBDLcBcaEUflBWshv2gAErZktT0tdEtXNRpv4vAvp0yEvAKSPVOESpnZW7PFNtBPI/+GlaAWxEC9V58qzhiRTJ+MU3fzwcBMRz4DmptdSN6bDLvkPr5eS9JQIDAQAB', 'token-service': 'http://localhost:8080/auth/realms/master/protocol/openid-connect', 'account-service': 'http://localhost:8080/auth/realms/master/account', 'tokens-not-before': 0}) http://localhost:8080/auth/realms/master/.well-known/openid-configuration
[INFO] Start scanner ClientScanner...
[INFO] Find a client for realm master: account
[INFO] Start scanner LoginScanner...
[+] LoginScanner - Form login work for admin on realm master, client account, (http://localhost:8080/auth/realms/master/account?session_state=4c152780-3980-439c-8e9d-15139ee19afa&code=2f821574-34e0-4917-8b00-c87c6fd302b0.4c152780-3980-439c-8e9d-15139ee19afa.3e118dc6-4780-42cf-90e7-abd81c1e7046)
[+] LoginScanner - Form login work for admin on realm master, client account, (http://localhost:8080/auth/realms/master/account?session_state=14a6cbcc-1b76-4b52-aa27-982a06b8c2a1&code=656fc3c9-3ea6-44af-9037-85288f471ab7.14a6cbcc-1b76-4b52-aa27-982a06b8c2a1.3e118dc6-4780-42cf-90e7-abd81c1e7046)
[INFO] Start scanner SecurityConsoleScanner...
[WARN] Result of SecurityConsoleScanner as no results (void list), subsequent scans can be void too.
[INFO] Start scanner OpenRedirectScanner...
[INFO] Start scanner FormPostXssScanner...
[INFO] Start scanner NoneSignScanner...
```

## scan types :

* realm : check if a realm exists
* client : check if a client exists in all realms
* well_known : get well_known for all realms
* login : test login against all clients / realms 
* client registration : try to add a new client (WARNING, client is deleted after test, if its not the case, be sure to make it manually)
* OpenRedirect : check if attack authorization flow via open redirection (unvalidated redirect_uri) is possible
* form post : check CVE 2018 14655 
* none sign : check if none sign algorithm is supported


## Help

```
$ keycloak-scanner --help
usage: keycloak-scanner [-h] --realms REALMS --clients CLIENTS [--proxy PROXY]

                        [--username USERNAME] [--password PASSWORD]

                        [--ssl-noverify] [--verbose] [--no-fail] [--fail-fast]

                        [--version]

                        (--registration-callback REGISTRATION_CALLBACK | --registration-callback-list REGISTRATION_CALLBACK_LIST)

                        base_url

KeyCloak vulnerabilities scanner.

positional arguments:

  base_url              URL to scan. ex http://localhost:8080

optional arguments:

  -h, --help            show this help message and exit

  --realms REALMS       Comma separated list of custom realms to test. ie :

                        master

  --clients CLIENTS     Comma separated list of custom clients to test. On

                        default installation, use account,admin-

                        cli,broker,realm-management,security-admin-console

  --proxy PROXY         Use a great proxy like BURP ;)

  --username USERNAME   If a username is specified, try to connect and attack

                        a token. If no password, try username as password.

  --password PASSWORD   password to test with username

  --ssl-noverify        Do not verify ssl certificates

  --verbose             Verbose mode

  --no-fail             Always exit with code 0 (by default, fail with an exit

                        code 4 if a vulnerability is discovered or 8 if an

                        error occur). Do NOT fail before all test are done.

  --fail-fast           Fail immediately if an error occur.

  --version             show program's version number and exit

  --registration-callback REGISTRATION_CALLBACK

                        Callback url to use on client registration test

  --registration-callback-list REGISTRATION_CALLBACK_LIST

                        File with one callback to test for registration by

                        line

Scans : 

- list realms

- Search well-known files

- Search for clients

- Search for valid logins

- Try client registration

- Search for security-admin-console and secret inside

- Search for open redirect via unvalidated redirect_uri

- Search for CVE-2018-14655 (reflected XSS)

- None alg in refresh token

Bugs, feature requests, request another scan, questions : https://github.com/NeuronAddict/keycloak-scanner.

*** Use it on production systems at your own risk ***

```


## Install with source code

With venv:

```
cd keycloak-scanner
python3 -m venv venv
source venv/bin/activate
pip install -e . # with -e, git pull will update code
keycloak-scanner
```

Or without venv :
```
cd keycloak-scanner
sudo pip3 install . # use sudo for install for all users
keycloak-scanner
```


## TODO

- password dictionary support
- Scanner details via command line
- 


CHANGELOG.md
# changelog

# version 1.0.4

- fail fast mode
- Examples in doc
- Don't display all objects on logs for better readability
- client registration scanner
- improve client scanner
- callback list support for client registration scanner
- delete client after client registration scan

# version 1.1.0

- Use a mediator pattern for more versatile scanners
- Fail fast mode
- Client registration scanner
- Improve command line options and default values
- Various bug fixes

# version 1.0.3

- Add login scanner
- Fix bad requests with Client(...) or Realm(...) in queries
- Fix proxy option
- Send warn in stderr
- Fix various bugs
- Better log when find something
- Provide version (--version option)

## Version 1.0.2

- hotfix : issue #20

## Version 1.0.1

- Script available in pip (pip install keycloak-scanner)
- Fix error on verbose mode
- Support for scan dependencies
- Return code 4 when vuln, 8 when no vuln but errors
- Fix bugs

## version 0.2.0

- Use fixed set of dependencies
- Improve tests
- Improve logs
- Add more tests
- fix issue #3
- Add travis deployment


keycloak_scanner/main.py
import argparse
import sys

import requests
import urllib3

from keycloak_scanner.scan_base.types import Username, Password
from keycloak_scanner.scan_base.wrap import WrapperTypes
from keycloak_scanner.scanners.clientregistration_scanner import ClientRegistrationScanner
from keycloak_scanner.scanners.clients_scanner import ClientScanner
from keycloak_scanner.scanners.form_post_xss_scanner import FormPostXssScanner
from keycloak_scanner.scanners.login_scanner import LoginScanner
from keycloak_scanner.scanners.none_sign_scanner import NoneSignScanner
from keycloak_scanner.scanners.open_redirect_scanner import OpenRedirectScanner
from keycloak_scanner.scanners.realm_scanner import RealmScanner
from keycloak_scanner.masterscanner import MasterScanner
from keycloak_scanner.scanners.security_console_scanner import SecurityConsoleScanner
from keycloak_scanner.scan_base.session_holder import SessionProvider
from keycloak_scanner.scanners.well_known_scanner import WellKnownScanner
from keycloak_scanner._version import __version__


def parser():
    parser = argparse.ArgumentParser(description='KeyCloak vulnerabilities scanner.',
                                     epilog='''

Scans : 
- list realms
- Search well-known files
- Search for clients
- Search for valid logins
- Try client registration
- Search for security-admin-console and secret inside
- Search for open redirect via unvalidated redirect_uri
- Search for CVE-2018-14655 (reflected XSS)
- None alg in refresh token

Bugs, feature requests, request another scan, questions : https://github.com/NeuronAddict/keycloak-scanner.

*** Use it on production systems at your own risk ***
''',
                                     formatter_class=argparse.RawDescriptionHelpFormatter)

    parser.add_argument('base_url', help='URL to scan. ex http://localhost:8080')
    parser.add_argument('--realms', help='Comma separated list of custom realms to test. Default: master', default='master')
    parser.add_argument('--clients', help='Comma separated list of custom clients to test. Default: account,admin-cli,broker,realm-management,security-admin-console',
                        default='account,admin-cli,broker,realm-management,security-admin-console')
    parser.add_argument('--proxy', help='Use an http proxy. ex: http://127.0.0.1:8080')
    parser.add_argument('--username', help='If a username is specified, try to connect and attack a token. If no '
                                           'password, try username as password.')
    parser.add_argument('--password', help='password to test with username')
    parser.add_argument('--ssl-noverify', help='Do not verify ssl certificates', action='store_true')
    parser.add_argument('--verbose', help='Verbose mode', action='store_true')
    parser.add_argument('--no-fail', action='store_true',
                        help='Always exit with code 0 (by default, fail with an exit code 4 if a vulnerability is '
                             'discovered or 8 if an error occur). '
                             'Do NOT fail before all test are done.')
    parser.add_argument('--fail-fast', action='store_true', help='Fail immediately if an error occur.')
    parser.add_argument('--version', action='version', version=f'keycloak-scanner {__version__}. '
                                                               f'https://github.com/NeuronAddict/keycloak-scanner.')

    group = parser.add_mutually_exclusive_group()
    group.add_argument('--registration-callback', help='Callback url to use on client registration test',
                       default='http://localhost')
    group.add_argument('--registration-callback-list', help='File with one callback to test for registration by line')

    return parser


def main():

    args = parser().parse_args()

    start(args, lambda: requests.Session())


def start(args, initial_session_provider: SessionProvider):

    realms = args.realms.split(',') if args.realms else []
    clients = args.clients.split(',') if args.clients else []

    def session_provider() -> requests.session():

        session = initial_session_provider()

        if args.proxy:
            session.proxies = {'http': args.proxy, 'https': args.proxy}

        if args.ssl_noverify:
            session.verify = False
            urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

        return session

    common_args = {
        'base_url': args.base_url,
        'verbose': args.verbose,
        'session_provider': session_provider
    }

    scanner = MasterScanner(scanners=[
        RealmScanner(realms=realms, **common_args),
        WellKnownScanner(**common_args),
        ClientScanner(clients=clients, **common_args),
        LoginScanner(**common_args),
        ClientRegistrationScanner(**common_args,
                                  callback_url=[args.registration_callback] if args.registration_callback else args.registration_callback_list), # TODO: list + file when str can confuse
        SecurityConsoleScanner(**common_args),
        OpenRedirectScanner(**common_args),
        FormPostXssScanner(**common_args),
        NoneSignScanner(**common_args)
    ], initial_values={
        WrapperTypes.USERNAME_TYPE: [Username(args.username)],
        WrapperTypes.PASSWORD_TYPE: [Password(args.password)],
    }, verbose=args.verbose, fail_fast=True)
    status = scanner.start()

    if not args.no_fail and status.has_vulns:
        print('Fail with exit code 4 because vulnerabilities are discovered')
        sys.exit(4)

    if status.has_error:
        print('No vulns but error(s) are returned, exit with code 8')
        sys.exit(8)

